### PR TITLE
Data science v1- hound issues fixed

### PR DESCRIPTION
**Pull request by ckhadilkar**

### CODE REVIEW

File: data_science/lib/cohort.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    @conversions = 0
+    @non_conversions = 0
+    process_data(data)
+  end
+
+  def process_data(data)
+    data.each do |record|
+      record["result"] == 1 ? @conversions += 1 : @non_conversions += 1
+    end
+  end
```
Comment (zsyed1) [inline data_science/lib/cohort.rb]: Your code is concise. An alternative way to do this is:
```
@conversions = data.select { |r| r["result"] == 1 }.size
@non_conversions = data.select { |r| r["result"] != 1 }.size
```

Comment (zsyed1) [inline data_science/lib/cohort.rb]: Or `@non_conversions = data.size - @conversions.size`

File: data_science/lib/cohort.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @conversions + @non_conversions
+  end
+
+  def conversion_rate
+    rate = conversions.to_f / total_sample_size
+    rate.round(DECIMAL_PLACES)
+  end
+
+  def std_err(p)
+    se = Math.sqrt(p * (1 - p) / total_sample_size)
```
Comment (zsyed1) [inline data_science/lib/cohort.rb]: I'm being real nit-picky here: `standard_error` is a better name.

File: data_science/lib/cohort.rb (hunk 1 above)
Comment (ckhadilkar) [inline data_science/lib/cohort.rb]: @conversions = data.select { |r| r["result"] == 1 }.size. I really like this one. Thanks. I ll definitely use it in future

File: data_science/lib/data_analyzer.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def create_cohorts
+    @cohort_a = Cohort.new(@data_parser.cohort_a)
+    @cohort_b = Cohort.new(@data_parser.cohort_b)
+  end
+
+  def build_ab_analyzer_data
+    data = {}
+    data[:acohort] = { success: @cohort_a.conversions, fails: @cohort_a.non_conversions }
```
Comment (zsyed1) [inline data_science/lib/data_analyzer.rb]: Why not use `data[:cohort_a]` instead of `data[:acohort]`?

File: data_science/lib/data_analyzer.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    data[:bcohort] = { success: @cohort_b.conversions, fails: @cohort_b.non_conversions }
+    data
```
Comment (zsyed1) [inline data_science/lib/data_analyzer.rb]: No need for the local variable `data`. Just do:
```
def build_ab_analyzer
  {
    cohort_a: {....},
    cohort_b: {....}
  }
end
```

File: data_science/lib/data_analyzer.rb (hunk 3 above)
Comment (ckhadilkar) [inline data_science/lib/data_analyzer.rb]: dont know why i did that :)

File: data_science/lib/data_analyzer.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    (1 - ab_tester.chisquare_p).round(DECIMAL_PLACES)
+  end
+
+  def display_data
+    puts "Cohort A"
+    puts @cohort_a
+    puts "Cohort B"
+    puts @cohort_b
+    puts "Confidence Level: #{calculate_confidence_level * 100}%"
+  end
```
Comment (zsyed1) [inline data_science/lib/data_analyzer.rb]: You can DRY the `puts` by interpolating: `puts "Cohort A\n#{@cohort_a}\nCohort B\n#{@cohort_b}"`

File: data_science/lib/data_parser.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    data_set = JSON.parse(data)
+    build_cohorts(data_set)
+    validate_cohort_data_size
+  end
+
+  def build_cohorts(data_set)
+    @cohort_a = data_set.select { |data| data["cohort"] == "A" }
+    @cohort_b = data_set.select { |data| data["cohort"] == "B" }
+  end
+
```
Comment (zsyed1) [inline data_science/lib/data_parser.rb]: You can do:
```
def build_cohorts(data_set, name)
  data_set.select { |data| data["cohort"] == name }
end
```
Then remove line 10 and change lines 6 and 7 to:
```
  @cohort_a = build_cohorts(data_set, "A")
  @cohort_b = build_cohorts(data_set, "B")
```

File: data_science/spec/data_parser_rspec.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  it 'raises error if file does not exist' do
+    expect { DataParser.new(FILE_NO_EXIST) }.to raise_error
+  end
+
+  it 'raises error if insufficient data' do
+    expect { DataParser.new(TEST_FILE_INSUFFICIENT_DATA) }.to raise_error
+  end
+end
```
Comment (zsyed1) [inline data_science/spec/data_parser_rspec.rb]: Great job! Do not think this another PR.

File: data_science/lib/cohort.rb (hunk 1 above)
Comment (djkotowski) [inline data_science/lib/cohort.rb]: No need for select + size.

```ruby
@conversions = data.count { |r| r["result"] == 1 }
@non_conversions = data.size - @conversions
```

File: data_science/lib/cohort.rb (hunk 2 above)
Comment (djkotowski) [inline data_science/lib/cohort.rb]: That's not super nit-picky.  Avoid small, unclear acronyms.

File: data_science/lib/data_analyzer.rb (hunk 5 above)
Comment (djkotowski) [inline data_science/lib/data_analyzer.rb]: Should the analyzer really be in charge of formatting data?

File: data_science/lib/data_parser.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  def validate_cohort_data_size
+    raise 'Insufficient data' unless @cohort_a.length >= MIN_DATA_SIZE ||
```
Comment (djkotowski) [inline data_science/lib/data_parser.rb]: Is a generic `RuntimeError` the best thing to raise here?  _(No need to change.  Just something to think about.)_

File: data_science/spec/cohort_rspec.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
+require_relative "spec_helper.rb"
+describe Cohort do
+  TEST_FILE = 'test_data_b.json'
+  let(:data_parser) { DataParser.new(TEST_FILE) }
+  let(:cohort_a) { Cohort.new(data_parser.cohort_a) }
+  let(:cohort_b) { Cohort.new(data_parser.cohort_b) }
+
+  it 'returns number of conversions for cohort A' do
+    expect(cohort_a.conversions).to eq(21)
+  end
```
Comment (djkotowski) [inline data_science/spec/cohort_rspec.rb]: You're describing `Cohort` here.  Cohort doesn't know whether it's cohort A or cohort B, so why should the tests know?

Also, look at using `describe "#method"` syntax to organize your tests.

File: data_science/spec/data_analyzer_rspec.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+require_relative "spec_helper.rb"
+describe DataAnalyzer do
+  TEST_FILE = 'test_data_b.json'
+  let(:data_analyzer) { DataAnalyzer.new(TEST_FILE) }
+
+  it 'initializes successfully' do
+    expect { data_analyzer }.not_to raise_error
+  end
+  it 'calculates and returns confidence leveli of 0.96' do
```
Comment (djkotowski) [inline data_science/spec/data_analyzer_rspec.rb]: Typo

File: data_science/lib/data_parser.rb (hunk 8 above)
Comment (ckhadilkar) [inline data_science/lib/data_parser.rb]: you mean, it should more specific text?

File: data_science/spec/data_parser_rspec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+require_relative "spec_helper.rb"
+describe DataParser do
+  TEST_FILE = 'test_data_b.json'
+  TEST_FILE_INSUFFICIENT_DATA = 'test_data_insufficient.json'
+  FILE_NO_EXIST = 'blah.json'
+
+  it 'reads input file successfully 'do
+    expect(DataParser.new(TEST_FILE)).to_not be_nil
```
Comment (djkotowski) [inline data_science/spec/data_parser_rspec.rb]: Prefer `not_to` to `to_not`.

File: data_science/spec/spec_helper.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
+require '../lib/cohort.rb'
+require '../lib/data_analyzer.rb'
+require '../lib/data_parser.rb'
+# This file was generated by the `rspec --init` command. Conventionally, all
```
Comment (djkotowski) [inline data_science/spec/spec_helper.rb]: Unless it's a file that is generated more than just once, delete generated comments.

File: data_science/lib/data_analyzer.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
+require '../lib/cohort.rb'
+require '../lib/data_parser.rb'
+require 'abanalyzer'
+class DataAnalyzer
+  DECIMAL_PLACES = 2
+  attr_accessor :data_parser
```
Comment (djkotowski) [inline data_science/lib/data_analyzer.rb]: Do you need a public accessor here?

File: data_science/lib/data_analyzer.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  def initialize(file)
+    @data_parser = DataParser.new(file)
+    create_cohorts
+  end
+
+  def create_cohorts
```
Comment (djkotowski) [inline data_science/lib/data_analyzer.rb]: Does this need to be public?

Comment (ckhadilkar) [inline data_science/lib/data_analyzer.rb]: Not really. i ll change it to private.

**Comment by djkotowski** [comment]
With DataAnalyzer, look at which methods you may be able to make `private`.  Otherwise, this is looking *really* good.  No need for another submission.  Just look at the comments for things to consider next time.

File: data_science/lib/data_parser.rb (hunk 8 above)
Comment (djkotowski) [inline data_science/lib/data_parser.rb]: I mean you might find another error class that better fits the issue.  Maybe `ArgumentError`.  Maybe a custom one.

File: data_science/spec/cohort_rspec.rb (hunk 9 above)
Comment (ckhadilkar) [inline data_science/spec/cohort_rspec.rb]: I can have a label that tells Cohort what its type is. Then have the test expect that. Because I do want to test calculations for both cohorts. 
Any other suggestion?

Comment (djkotowski) [inline data_science/spec/cohort_rspec.rb]: Test both cohorts, but because you just are verifying that the functionality is the same for two cohorts, merge them into the same test.  This is a case where two expectations in the same test work.
